File: src/types/analysis.ts
import type { Header } from './core'

export type TokenKind =
  | 'bearer'
  | 'jwt'
  | 'api-key'
  | 'csrf'
  | 'session-id'
  | 'request-id'
  | 'basic'
  | 'unknown'

export type TokenSource = 'header' | 'cookie' | 'query' | 'body'

export type LinkKind =
  | 'cookie'
  | 'header-value'
  | 'query-value'
  | 'body-value'
  | 'json-value'
  | 'path-value'

export interface SessionAnalysis {
  requests: number
  domains: number
  api_endpoints: number
  unique_endpoints: number
  json_responses: number
  post_requests: number
  with_cookies: number
  possible_tokens: number
  high_importance: number
  errors: number
  total_bytes: number
}

export interface EndpointGroup {
  normalized: string
  host: string
  methods: string[]
  count: number
  is_api: boolean
  sample_request_id: string
  sequence_ids: number[]
  status_codes: number[]
  avg_duration_ms: number
}

export interface CookieEvent {
  request_id: string
  sequence_id: number
  method: string
  path: string
  value_preview: string
}

export interface CookieUsage {
  name: string
  domain: string
  value_preview: string
  distinct_values: number
  created_by: CookieEvent[]
  used_by: CookieEvent[]
}

export interface DetectedToken {
  kind: TokenKind
  source: TokenSource
  name: string
  value_preview: string
  value_hash: string
  used_by: number[]
  first_seen_request_id: string
}

export interface Relationship {
  from_request_id: string
  from_sequence_id: number
  from_path: string
  to_request_id: string
  to_sequence_id: number
  to_path: string
  kind: LinkKind
  value_preview: string
  source_json_path: string | null
  target_location: string
  confidence: number
}

export interface FlowNode {
  id: string
  label: string
  host: string
  method: string
  count: number
  importance: string
  depth: number
  sample_request_id: string
}

export interface FlowEdge {
  from: string
  to: string
  kind: LinkKind
  label: string
  weight: number
}

export interface FlowGraph {
  nodes: FlowNode[]
  edges: FlowEdge[]
}

export interface AnalysisBundle {
  session_id: string
  generated_at: number
  overview: SessionAnalysis
  endpoints: EndpointGroup[]
  tokens: DetectedToken[]
  relationships: Relationship[]
  graph: FlowGraph
  cookies: CookieUsage[]
  truncated: boolean
}

export interface DetectedId {
  value: string
  location: string
  kind: string
}

export interface RequestAnalysis {
  importance: string
  reasons: string[]
  normalized_endpoint: string
  is_api: boolean
  detected_ids: DetectedId[]
  tokens: DetectedToken[]
  inbound: Relationship[]
  outbound: Relationship[]
  repeat_count: number
}

export interface CaInfo {
  exists: boolean
  common_name: string
  cert_path: string
  key_path: string
  fingerprint: string | null
  not_after: string | null
  installed: boolean
}

export interface TrustStorePlan {
  platform: string
  steps: string[]
  requires_elevation: boolean
  manual_instructions: string[]
}

export type BrowserKind = 'chromium' | 'firefox'

export interface BrowserOption {
  id: string
  name: string
  path: string
  kind: BrowserKind
  uses_system_trust: boolean
}

export interface AppSettings {
  language: string
  proxy_port: number
  mask_secrets: boolean
  timeline_page_size: number
  auto_analyse: boolean
  default_replay_delay_ms: number
  sensitive_headers: string[]
}

export interface StorageInfo {
  data_dir: string
  database_bytes: number
  blob_bytes: number
}

export interface ImportReport {
  session_id: string
  imported: number
  skipped: number
}

export interface ComparisonSide {
  label: string
  status: number | null
  duration_ms: number
  size: number
  content_type: string | null
  headers: Header[]
  body: string | null
}

export interface DiffEntry {
  kind: 'added' | 'removed' | 'changed'
  path: string
  left: string | null
  right: string | null
  volatile: boolean
}

export interface Comparison {
  left: ComparisonSide
  right: ComparisonSide
  header_diff: DiffEntry[]
  body_diff: DiffEntry[]
  body_comparable: boolean
}

File: src-tauri/src/models/analysis.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionAnalysis {
    pub requests: i64,
    pub domains: i64,
    pub api_endpoints: i64,
    pub unique_endpoints: i64,
    pub json_responses: i64,
    pub post_requests: i64,
    pub with_cookies: i64,
    pub possible_tokens: i64,
    pub high_importance: i64,
    pub errors: i64,
    pub total_bytes: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EndpointGroup {
    pub normalized: String,
    pub host: String,
    pub methods: Vec<String>,
    pub count: i64,
    pub is_api: bool,
    pub sample_request_id: String,
    pub sequence_ids: Vec<i64>,
    pub status_codes: Vec<u16>,
    pub avg_duration_ms: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CookieUsage {
    pub name: String,
    pub domain: String,
    pub value_preview: String,
    pub distinct_values: i64,
    pub created_by: Vec<CookieEvent>,
    pub used_by: Vec<CookieEvent>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CookieEvent {
    pub request_id: String,
    pub sequence_id: i64,
    pub method: String,
    pub path: String,
    pub value_preview: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum TokenKind {
    Bearer,
    Jwt,
    ApiKey,
    Csrf,
    SessionId,
    RequestId,
    Basic,
    Unknown,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum TokenSource {
    Header,
    Cookie,
    Query,
    Body,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetectedToken {
    pub kind: TokenKind,
    pub source: TokenSource,
    pub name: String,
    pub value_preview: String,
    pub value_hash: String,
    pub used_by: Vec<i64>,
    pub first_seen_request_id: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum LinkKind {
    Cookie,
    HeaderValue,
    QueryValue,
    BodyValue,
    JsonValue,
    PathValue,
}

/// A heuristic link: a value produced by one exchange reappearing in a later
/// request. Suggestive, never a guarantee.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Relationship {
    pub from_request_id: String,
    pub from_sequence_id: i64,
    pub from_path: String,
    pub to_request_id: String,
    pub to_sequence_id: i64,
    pub to_path: String,
    pub kind: LinkKind,
    pub value_preview: String,
    pub source_json_path: Option<String>,
    pub target_location: String,
    pub confidence: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FlowNode {
    pub id: String,
    pub label: String,
    pub host: String,
    pub method: String,
    pub count: i64,
    pub importance: String,
    pub depth: i32,
    pub sample_request_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FlowEdge {
    pub from: String,
    pub to: String,
    pub kind: LinkKind,
    pub label: String,
    pub weight: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FlowGraph {
    pub nodes: Vec<FlowNode>,
    pub edges: Vec<FlowEdge>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestAnalysis {
    pub importance: String,
    pub reasons: Vec<String>,
    pub normalized_endpoint: String,
    pub is_api: bool,
    pub detected_ids: Vec<DetectedId>,
    pub tokens: Vec<DetectedToken>,
    pub inbound: Vec<Relationship>,
    pub outbound: Vec<Relationship>,
    pub repeat_count: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetectedId {
    pub value: String,
    pub location: String,
    pub kind: String,
}
